### ISSUE TITLE
LXC Builder + Ansible Remote = hung shell

### ISSUE THREAD

**Issue by conn** [hashicorp/packer#5485]
Hello!

I'm pretty excited to see that an LXC builder has made it into v1.1.1. Sadly, while shell provisioning works with it, ansible provisioning currently does not. Basically what happens is that the first command  ran through the LXC communicator hangs forever. I have a gist [here](https://gist.github.com/conn/9717d0909b08e57802b7d1b58d1fbd88) with all pertinent information.

I've been throwing myself at this problem for a few days and can't seem to figure out the cause. Raw ssh commands through the communicator do just fine and I'm even able to get an interactive shell while ansible is hanging. Ansible also works just fine when ran with other builders such as qemu.

There's some weird interaction between the LXC communicator and ansible that's causing this.

Please let me know if you have any insight.

**Comment by rickard-von-essen** [comment]
See my answer on the [mailing list](https://groups.google.com/forum/#!msg/packer-tool/sJqMu8DgjZM/szjiWtXFBAAJ). I'll close this since it should work if you setup Ansible correctly and we can reopen if that is not the case.

**Comment by conn** [comment]
@rickard-von-essen 

Did you compare output between the two? These are completely different issues. The one in the mailing list is dealing with not being able to connect via SSH to run any command. I specifically mentioned that I was able to connect through the SSH communicator just fine:
```
root@vagrant:~# ssh -p 35353 -i /tmp/ansible-key408993886  127.0.0.1 -- /bin/bash -i

root@packer-chroot:/# hostname
hostname
packer-chroot

root@packer-chroot:/# echo "$container"
lxc
echo "$container"
```
The playbook stops after connecting successfully:
```
2017/10/19 22:52:20 packer: 2017/10/19 22:52:20 SSH proxy: accepted connection
2017/10/19 22:52:20 packer: 2017/10/19 22:52:20 authentication attempt from 10.0.0.10:48344 to 127.0.0.1:43413 as root using none
2017/10/19 22:52:20 packer: 2017/10/19 22:52:20 authentication attempt from 10.0.0.10:48344 to 127.0.0.1:43413 as root using publickey
2017/10/19 22:52:20 packer: 2017/10/19 22:52:20 new exec request: /bin/sh -c 'echo ~ && sleep 0'
2017/10/19 22:52:20 packer: 2017/10/19 22:52:20 Executing with lxc-attach in container: packer-chroot /var/lib/lxc/packer-chroot/rootfs /bin/sh -c 'echo ~ && sleep 0'
2017/10/19 22:52:20 packer: 2017/10/19 22:52:20 Executing lxc-attach: /bin/sh []string{"/bin/sh", "-c", "sudo lxc-attach --name packer-chroot -- /bin/sh -c \"/bin/sh -c 'echo ~ && sleep 0'\""}
```

**Comment by rickard-von-essen** [comment]
Yes and I still think they are the same. Packer don't use SSH to communicate with LXC, but `lxc-attach`.

**Comment by conn** [comment]
Due to comments [here](https://github.com/hashicorp/packer/issues/3331) I would think that I should not have to worry about the connection details and let the communicator abstraction take care of it for me?

**Comment by conn** [comment]
Also, what is the point of having Ansible's SSH proxy connect with LXC's communicator if it was never meant to do so? And why is this caveat not mentioned in the [documentation](https://www.packer.io/docs/builders/lxc.html)?

**Comment by rickard-von-essen** [comment]
> I would think that I should not have to worry about the connection details and let the communicator abstraction take care of it for me?

Unfortunately not for the Ansible provisioner. The communicator type is explicitly hidden for the provisioner.

> Ansible remote provisioner's SSH proxy connect with LXC's communicator if it was never meant to do so?

Because of how Ansible executes from a remote host which doesn't fit very well into the design of provisioners in Packer. Reading through previously mentioned #3331 and linked issues gives a good background and reasoning.

> And why is this caveat not mentioned in the documentation?

Because it's new and community maintained, but it doesn't explicitly say _"In addition to the options listed here, a communicator can be configured for this builder."_

It would be great if you could contribute a working example to add to the [Ansible Remote documentation](https://www.packer.io/docs/provisioners/ansible.html#chroot-communicator).

**Comment by conn** [comment]
Just to reiterate words in #3331:

> I appreciate the flexibility here but I can't support this. This is not packer anymore. Constraints are useful.
> Also, the communicator may not be communicating using the TCP/IP at all; the communicator abstraction is useful precisely because it defines a contract that allows clients to be unconcerned about the transport.

There seems to be a lack of consistency toward what Packer is and isn't responsible for:
In one place, the maintainers repeatedly refuse to provide needed tools for the users to empower themselves simply because it goes against what Packer should be; in another place the maintainers refuse to acknowledge that same idea and ignore bugs that break the mentioned "contract", leaving the users to hack around it themselves.

I'll still try the workaround and contribute documentation if I'm successful, but I ask that you reconsider this issue to be more than documentation-related.

**Comment by rickard-von-essen** [comment]
You are kind of shooting the messenger. But until we change our standpoint of the connection abstraction I still think there's only some missing documentation.

**Comment by SwampDragons** [comment]
@conn thanks for the report. I think you have a good point here; If there's a way to handle the ansible plugins ourselves it works effortlessly for users without them having to manually configure something outside Packer, I'm all for it.  The ansible provisioner itself is a community-maintained provisioner, so I'd gladly review a pull request that introduced this behavior.

In the meantime, please do let us know if configuring the plugin allows you to use the ansible provisioner with the lxc builder.  While it may not match our philosophy about abstracting away the communication layer (which I think is a fair criticism), at least if it works with the plugin it will be consistent with how ansible behaves for other non-ssh communicators.

**Comment by conn** [comment]
@rickard-von-essen I'm sorry if you feel I'm attacking you. Being on the receiving end of this bug, and also a bug that might be easily solved by the variables requested in #3331, I find it frustrating to receive two conflicting messages:

> We're not going to give you variables to empower yourself simply because this constraint is part of our model
> Instead of working this bug into our model, we're going to disregard and just publish a workaround as documentation

Please don't view this as a personal attack on you, but just more of an overall observation toward how these bugs are being handled here (and also other Hashicorp projects). In the least, I would expect for the issue to be left open for the community to pick up and as a reminder to work toward a better model; whether you are the messenger or not, being dismissive about a break from expected behavior like this isn't cool and shuts down the process toward quality contributions. 

@SwampDragons I don't really have a fix at the moment and probably won't for a while. I'm all for getting things working `$now` and if I'm able to do so, I'll be happy to contribute documentation. That being said, the LXC connector for Ansible isn't work too great (it doesn't seem too mature) so I have to go chasing bugs there. I`ll probably work around it by starting SSH in the container one way or another and then provisioning over it.

**Comment by SwampDragons** [comment]
Have you opened an issue for the bug that would be solved by having access to the variables in #3331?

**Comment by conn** [comment]
@SwampDragons It's related to work so it's not really up to me to disclose all the details. Basically, there are some subtle behavior changes when you add a proxy into the mix as you provision with Ansible. One such instance is when you're rebooting (after installing a new version of the kernel via a role) and waiting for the box (Openstack instance) to come back up, you'll get handshake failures even  when using `wait_for`. This is something that only happens with the Ansible remote provisioner + the Openstack builder. While this normally would not make the playbook fail, we're using build scripts that parse Ansible's output for errors and the handshake failure trips it up.

In short, it would be _really_ nice to be able to pass the instance's IP directly to Ansible to get around the proxy altogether. I know there are more elegant ways to work around this, but having that connection information available would be the easiest and fastest way from A -> B instead of going down yet another rabbit hole.

**Comment by conn** [comment]
I've raised an [issue](https://github.com/ansible/ansible/issues/32005) with Ansible for a bug with the LXC connector. It's a bit of a blocker.

**Comment by SwampDragons** [comment]
We've recently added the variables requested in #3331 -- take a look at the docs for the `build` template engine [here](https://packer.io/docs/templates/engine.html#build). Hopefully this un-sticks you.
